File: backend/src/models/Comment.js
const mongoose = require('mongoose');

const CommentSchema = new mongoose.Schema({
  contentId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Content',
    required: true,
    index: true
  },
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  
  text: {
    type: String,
    required: true,
    maxlength: 500
  },
  
  // Threading
  parentId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Comment',
    index: true
  },
  
  // Mentions
  mentions: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  
  // Engagement
  likesCount: {
    type: Number,
    default: 0,
    min: 0
  },
  repliesCount: {
    type: Number,
    default: 0,
    min: 0
  },
  
  // Moderation
  isHidden: {
    type: Boolean,
    default: false
  },
  isReported: {
    type: Boolean,
    default: false
  },
  reportCount: {
    type: Number,
    default: 0
  },
  
  // Pinned by creator
  isPinned: {
    type: Boolean,
    default: false
  },
  
  isEdited: {
    type: Boolean,
    default: false
  },
  editedAt: Date,
  
  deletedAt: Date,
  
}, {
  timestamps: true
});

// Compound indexes
CommentSchema.index({ contentId: 1, createdAt: -1 });
CommentSchema.index({ userId: 1, createdAt: -1 });
CommentSchema.index({ parentId: 1, createdAt: -1 });
CommentSchema.index({ contentId: 1, isPinned: -1, likesCount: -1 });

// Static method to get top comments
CommentSchema.statics.getTopComments = function(contentId, limit = 10) {
  return this.find({
    contentId,
    parentId: null, // Only top-level comments
    isHidden: false
  })
  .sort({ isPinned: -1, likesCount: -1, createdAt: -1 })
  .limit(limit)
  .populate('userId', 'username fullName avatar isVerified');
};

// Static method to get replies
CommentSchema.statics.getReplies = function(parentId, limit = 20) {
  return this.find({
    parentId,
    isHidden: false
  })
  .sort({ createdAt: 1 })
  .limit(limit)
  .populate('userId', 'username fullName avatar isVerified');
};

const Comment = mongoose.model('Comment', CommentSchema);

module.exports = Comment;

File: flutter_app/lib/features/content/data/models/comment.dart
class Comment {
  final String id;
  final String contentId;
  final String userId;
  final String? username;
  final String? userAvatar;
  final String text;
  final int likes;
  final bool isLiked;
  final int replies;
  final String? parentCommentId;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Comment({
    required this.id,
    required this.contentId,
    required this.userId,
    this.username,
    this.userAvatar,
    required this.text,
    this.likes = 0,
    this.isLiked = false,
    this.replies = 0,
    this.parentCommentId,
    this.createdAt,
    this.updatedAt,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      id: json['_id'] ?? json['id'] ?? '',
      contentId: json['contentId'] ?? json['content'] ?? '',
      userId: json['userId'] ?? json['user']?['_id'] ?? '',
      username: json['username'] ?? json['user']?['username'],
      userAvatar: json['userAvatar'] ?? json['user']?['avatar'],
      text: json['text'] ?? '',
      likes: json['likes'] ?? json['likesCount'] ?? 0,
      isLiked: json['isLiked'] ?? false,
      replies: json['replies'] ?? json['repliesCount'] ?? 0,
      parentCommentId: json['parentCommentId'],
      createdAt:
          json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt:
          json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'contentId': contentId,
      'userId': userId,
      if (username != null) 'username': username,
      if (userAvatar != null) 'userAvatar': userAvatar,
      'text': text,
      'likes': likes,
      'isLiked': isLiked,
      'replies': replies,
      if (parentCommentId != null) 'parentCommentId': parentCommentId,
      if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updatedAt': updatedAt!.toIso8601String(),
    };
  }

  Comment copyWith({
    String? id,
    String? contentId,
    String? userId,
    String? username,
    String? userAvatar,
    String? text,
    int? likes,
    bool? isLiked,
    int? replies,
    String? parentCommentId,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Comment(
      id: id ?? this.id,
      contentId: contentId ?? this.contentId,
      userId: userId ?? this.userId,
      username: username ?? this.username,
      userAvatar: userAvatar ?? this.userAvatar,
      text: text ?? this.text,
      likes: likes ?? this.likes,
      isLiked: isLiked ?? this.isLiked,
      replies: replies ?? this.replies,
      parentCommentId: parentCommentId ?? this.parentCommentId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
